### PR TITLE
Add tree_labels, tree_print.

### PR DESCRIPTION
**Pull request by awf**
This PR follows from https://github.com/tensorflow/tensorflow/pull/56202, where the pytree walk function has been extended to pass node_data to walkers.  For example, this allows implementation of custom printers (see https://github.com/google/jax/issues/5047), which can e.g. print a tree with path labels like:

```
obj/a1 = leaf-a1
obj/a2/b1/0 = 111
obj/a2/b1/1 = 2
obj/a2/b1/2 = 3
obj/a2/b2 = [[11. 12. 13.]; [21. 22. 23.]]
obj/a2/b3/c1 = 33
obj/a2/b3/c2/0 = 44
obj/a2/b3/c2/1 = 55
obj/a3/x1/x11 = a
obj/a3/x1/x12/y1/y11 = ya
obj/a3/x1/x12/y1/y12 = yb
obj/a3/x1/x13 = b
obj/a3/x2/x21 = 2a
obj/a3/x2/x22/y1/y11 = ya
obj/a3/x2/x22/y1/y12 = yb
obj/a4 = fred
```

### CODE REVIEW

File: jax/_src/tree_util.py [hunk 1]
```diff
@@ -204,7 +204,7 @@ def tree_multimap(*args, **kwargs):
 # TODO(mattjj,phawkins): consider removing this function
 def _process_pytree(process_node, tree):
   leaves, treedef = pytree.flatten(tree)
-  return treedef.walk(process_node, None, leaves), treedef
+  return treedef.walk(lambda node, data: process_node(node), None, leaves), treedef
```
Comment (awf) [inline jax/_src/tree_util.py:208]: Adapt to old API, could just choose now to delete this function, or propagate the API to the currently unknown list of clients of this function.

File: jax/_src/tree_util.py [hunk 2]
```diff
@@ -22,7 +22,7 @@
 import textwrap
 import warnings
 
-from jax._src.lib import pytree
+from jax._src.lib import pytree, xla_extension
```
Comment (awf) [inline jax/_src/tree_util.py:25]: See TODO below.

File: tests/tree_util_test.py [hunk 3]
```diff
@@ -126,8 +132,8 @@ def tree_unflatten(cls, meta, data):
                              [("foo", 34), ("baz", 101), ("something", -42)]),),
     (ANamedTupleSubclass(foo="hello", bar=3.5),),
     (FlatCache(None),),
-    (FlatCache(1),),
-    (FlatCache({"a": [1, 2]}),),
+    (FlatCache(111),),
+    (FlatCache({"a": [111, 112]}),),
```
Comment (awf) [inline tests/tree_util_test.py:136]: This is just to make them visually more distinctive in debug output

File: tests/tree_util_test.py [hunk 4]
```diff
@@ -15,10 +15,16 @@
...
+import types
+import pprint
+
+import unittest
 
 from absl.testing import absltest
 from absl.testing import parameterized
 
+import jax
+
```
Comment (awf) [inline tests/tree_util_test.py:27]: Needed for xla_extension_version check below.

**Comment by awf** [comment]
@hawkinsp as you are looking at the corresponding TF PR, this is the JAX one.

**Comment by awf** [comment]
@hawkinsp I have rebased this on top of main - hopefully it's OK to approve given the TF change went in.

**Comment by awf** [comment]
@hawkinsp I have rebased this on top of main - hopefully it's OK to approve given the TF change went in.
